### DISCUSSION TITLE
IKE_UP_DOWN event behavior on Reauth operation

### DISCUSSION THREAD

**Discussion by hbhokare** [strongswan/strongswan#573] [Q&A]
Hi,
We are trying to develop the IPSec-based application and using different strongswan/davici API from our application. We have the following questions related to event handling during reauth operation. 

Case 1: make-before-break = yes 
Do we always expect ike up event first with a new unique id and then ike down event for existing ike sa with known unique id which is same as the first ike up unique id ? if yes then can we always believe in the fact that our application would get ike up for which we can allocate new ike sa obj in an application and then an application would get ike down on which application would deallocate the existing ike sa object corresponding to unique id received in ike down event. 

Case 2: make-before-break = no
Do we expect ike down first for existing ike sa and then would get new ike up event on which we need to allocate ike sa object in an application ? if yes then can we expect the secure tunnel break in this case

Thanks,
-Harish

**Comment by tobiasbrunner** [comment]
Yes to all of your questions.

**Reply by hbhokare** [reply]
Thanks !

I am trying to set make_before_break = yes using the following strongswan.conf but in the log, I can always see ike down first and then ike up which is like case 2 from the above scenario. Am I missing something how can I make sure that make_before_break policy is correctly set 

/etc/strongswan/strongswan.con
charon {
        load_modular = yes
        plugins {
                include strongswan.d/charon/*.conf
        }
  filelog {
        charon {
            # path to the log file, specify this as section name in versions prior to 5.7.0
            path = /dumps/charon.log
            # add a timestamp prefix
            time_format = %b %e %T
            # prepend connection name, simplifies grepping
            ike_name = yes
            # overwrite existing files
            append = yes
            # increase default loglevel for all daemon subsystems
            default = 2
            # flush each line to disk
            flush_line = yes
            make_before_break = yes
        }
stderr {
            # more detailed loglevel for a specific subsystem, overriding the
            # default loglevel.
            ike = 4
            knl = 4
            cfg = 4
            net = 4
        }
    }
}
include strongswan.d/*.conf

**Reply by tobiasbrunner** [reply]
You put that at the wrong location (inside the `charon.filelog.charon` section). Move it to the same level as e.g. `load_modular`.

**Reply by hbhokare** [reply]
Thanks! I am able to see new ike up first and then ike down for older ike sa but here are few observations.

I can also see another ike down event which is corresponding to new ike up event sent during reauth operation, after the older ike down event received and there is also another new ike up with different unique id.  is it expected ?

**Reply by hbhokare** [reply]
Note - I am triggering reauth operation using swanctl --rekey --reauth --ike connectionAB from site A
charon log for above issue -
Sep  1 08:22:03 12[CFG] vici client 4 requests: initiate
Sep  1 08:22:03 12[CFG] vici initiate IKE_SA 'connectionAB'
Sep  1 08:22:03 11[IKE] <connectionAB|1> nothing to initiate
Sep  1 08:22:13 05[CFG] vici client 5 requests: initiate
Sep  1 08:22:13 05[CFG] vici initiate CHILD_SA 'connectionAB'
Sep  1 08:22:13 10[IKE] <connectionAB|1> nothing to initiate
Sep  1 08:22:35 11[CFG] vici client 7 requests: rekey
Sep  1 08:22:35 11[CFG] vici rekey IKE_SA 'connectionAB'
Sep  1 08:22:35 11[MGR] checkout IKEv2 SA with SPIs 500eb5edaab53a79_i 3a453f9b335f3d65_r
Sep  1 08:22:35 11[MGR] IKE_SA connectionAB[1] successfully checked out
Sep  1 08:22:35 11[IKE] <connectionAB|1> reauthenticating IKE_SA connectionAB[1]
Sep  1 08:22:35 11[IKE] <connectionAB|1> queueing IKE_VERIFY_PEER_CERT task
Sep  1 08:22:35 11[IKE] <connectionAB|1> queueing IKE_REAUTH_COMPLETE task
Sep  1 08:22:35 11[MGR] <connectionAB|1> checkin IKE_SA connectionAB[2]
Sep  1 08:22:35 11[MGR] <connectionAB|1> checkin of IKE_SA successful
Sep  1 08:22:35 11[IKE] <connectionAB|1> activating new tasks
Sep  1 08:22:35 11[IKE] <connectionAB|1> nothing to initiate
Sep  1 08:22:35 11[MGR] <connectionAB|1> checkin IKE_SA connectionAB[1]
Sep  1 08:22:35 11[MGR] <connectionAB|1> checkin of IKE_SA successful
Sep  1 08:22:35 02[JOB] watched FD 24 ready to read
Sep  1 08:22:35 02[JOB] watcher going to poll() 9 fds
Sep  1 08:22:35 16[CFG] vici client 7 disconnected
Sep  1 08:22:35 02[JOB] watcher got notification, rebuilding
Sep  1 08:22:35 02[JOB] watcher going to poll() 9 fds
Sep  1 08:22:35 03[NET] received packet: from B[4500] to A[4500]
Sep  1 08:22:35 03[ENC] parsing header of message
Sep  1 08:22:35 03[ENC] parsing HEADER payload, 280 bytes left
Sep  1 08:22:35 03[ENC]   parsing rule 0 IKE_SPI
Sep  1 08:22:35 03[ENC]   parsing rule 1 IKE_SPI
Sep  1 08:22:35 03[ENC]   parsing rule 2 U_INT_8
Sep  1 08:22:35 03[ENC]   parsing rule 3 U_INT_4
Sep  1 08:22:35 03[ENC]   parsing rule 4 U_INT_4
Sep  1 08:22:35 03[ENC]   parsing rule 5 U_INT_8
Sep  1 08:22:35 03[ENC]   parsing rule 6 RESERVED_BIT
Sep  1 08:22:35 03[ENC]   parsing rule 7 RESERVED_BIT
Sep  1 08:22:35 03[ENC]   parsing rule 8 FLAG
Sep  1 08:22:35 03[ENC]   parsing rule 9 FLAG
Sep  1 08:22:35 03[ENC]   parsing rule 10 FLAG
Sep  1 08:22:35 03[ENC]   parsing rule 11 FLAG
Sep  1 08:22:35 03[ENC]   parsing rule 12 FLAG
Sep  1 08:22:35 03[ENC]   parsing rule 13 FLAG
Sep  1 08:22:35 03[ENC]   parsing rule 14 U_INT_32
Sep  1 08:22:35 03[ENC]   parsing rule 15 HEADER_LENGTH
Sep  1 08:22:35 03[ENC] parsing HEADER payload finished
Sep  1 08:22:35 03[ENC] parsed a IKE_SA_INIT response header
Sep  1 08:22:35 03[NET] waiting for data on sockets
Sep  1 08:22:35 06[MGR] checkout IKEv2 SA by message with SPIs 122e931ace2dd372_i dba7144218752388_r
Sep  1 08:22:35 06[MGR] IKE_SA connectionAB[2] successfully checked out
Sep  1 08:22:35 07[ENC] <connectionAB|1> parsing DELETE payload, 8 bytes left
Sep  1 08:22:35 07[ENC] <connectionAB|1> parsing DELETE payload finished
Sep  1 08:22:35 07[ENC] <connectionAB|1> insert decrypted payload of type DELETE at end of list
Sep  1 08:22:35 07[ENC] <connectionAB|1> found payload of type DELETE
Sep  1 08:22:35 07[IKE] <connectionAB|1> received DELETE for IKE_SA connectionAB[1]
Sep  1 08:22:35 07[IKE] <connectionAB|1> IKE_SA connectionAB[1] state change: DELETING => DESTROYING
Sep  1 08:22:35 07[CHD] <connectionAB|1> CHILD_SA connectionAB{1} state change: INSTALLED => DESTROYING
Sep  1 08:22:35 07[KNL] <connectionAB|1> deleting policy A/32 === B/32 out
Sep  1 08:22:35 07[KNL] <connectionAB|1> policy still used by another CHILD_SA, not removed
Sep  1 08:22:35 07[KNL] <connectionAB|1> not updating policy A/32 === B/32 out [priority 367231, refcount 1]
Sep  1 08:22:35 07[KNL] <connectionAB|1> deleting policy B/32 === A/32 in
Sep  1 08:22:35 07[KNL] <connectionAB|1> policy still used by another CHILD_SA, not removed
Sep  1 08:22:35 07[KNL] <connectionAB|1> not updating policy B/32 === A/32 in [priority 367231, refcount 1]
Sep  1 08:22:35 07[KNL] <connectionAB|1> deleting policy B/32 === A/32 fwd
Sep  1 08:22:35 07[KNL] <connectionAB|1> policy still used by another CHILD_SA, not removed
Sep  1 08:22:35 07[KNL] <connectionAB|1> not updating policy B/32 === A/32 fwd [priority 367231, refcount 1]
Sep  1 08:22:35 07[KNL] <connectionAB|1> deleting SAD entry with SPI c9e141dd
Sep  1 08:22:35 07[KNL] <connectionAB|1> deleted SAD entry with SPI c9e141dd
Sep  1 08:22:35 07[KNL] <connectionAB|1> deleting SAD entry with SPI c669827b
Sep  1 08:22:35 07[KNL] <connectionAB|1> deleted SAD entry with SPI c669827b
Sep  1 08:22:35 07[MGR] checkin and destroy of IKE_SA successful
Sep  1 08:22:35 10[ENC] <connectionAB|2> parsing DELETE payload, 8 bytes left
Sep  1 08:22:35 10[ENC] <connectionAB|2> parsing DELETE payload finished
Sep  1 08:22:35 10[ENC] <connectionAB|2> insert decrypted payload of type DELETE at end of list
Sep  1 08:22:35 10[ENC] <connectionAB|2> found payload of type DELETE
Sep  1 08:22:35 10[IKE] <connectionAB|2> received DELETE for IKE_SA connectionAB[2]
Sep  1 08:22:35 10[NET] connectionAB|2> sending packet: from A[4500] to B[4500] (80 bytes)
Sep  1 08:22:35 10[MGR] connectionAB|2> checkin and destroy IKE_SA connectionAB[2]
Sep  1 08:22:35 10[IKE] connectionAB|2> IKE_SA connectionAB[2] state change: DELETING => DESTROYING
Sep  1 08:22:35 10[CHD] connectionAB|2> CHILD_SA connectionAB{2} state change: INSTALLED => DESTROYING
Sep  1 08:22:35 10[KNL] connectionAB|2> deleting policy A/32 === B/32 out
Sep  1 08:22:35 10[KNL] connectionAB|2> deleting policy B/32 === A/32 in
Sep  1 08:22:35 10[KNL] connectionAB|2> deleting policy B/32 === A/32 fwd
Sep  1 08:22:35 10[KNL] connectionAB|2> deleting SAD entry with SPI caa1d755
Sep  1 08:22:35 10[KNL] connectionAB|2> deleted SAD entry with SPI caa1d755
Sep  1 08:22:35 10[KNL] connectionAB|2> deleting SAD entry with SPI c60b0f06
Sep  1 08:22:35 10[KNL] connectionAB|2> deleted SAD entry with SPI c60b0f06
Sep  1 08:22:35 10[MGR] checkin and destroy of IKE_SA successful
Sep  1 08:22:35 11[IKE] <connectionAB|1> queueing IKE_REAUTH_COMPLETE task
Sep  1 08:22:35 14[IKE] <connectionAB|2>   activating IKE_REAUTH_COMPLETE task
Sep  1 08:22:35 04[NET] sending packet: from A[4500] to B[4500]
Sep  1 08:22:35 02[JOB] watched FD 19 ready to read
Sep  1 08:22:35 02[JOB] watcher going to poll() 8 fds
Sep  1 08:22:35 02[JOB] watcher got notification, rebuilding
Sep  1 08:22:35 02[JOB] watcher going to poll() 9 fds
Sep  1 08:22:35 03[NET] received packet: from B[4500] to A[4500]
Sep  1 08:22:35 03[ENC] parsing header of message
Sep  1 08:22:35 03[ENC] parsing HEADER payload, 280 bytes left
Sep  1 08:22:35 03[ENC]   parsing rule 0 IKE_SPI
Sep  1 08:22:35 03[ENC]   parsing rule 1 IKE_SPI
Sep  1 08:22:35 03[ENC]   parsing rule 2 U_INT_8
Sep  1 08:22:35 03[ENC]   parsing rule 3 U_INT_4
Sep  1 08:22:35 03[ENC]   parsing rule 4 U_INT_4
Sep  1 08:22:35 03[ENC]   parsing rule 5 U_INT_8
Sep  1 08:22:35 03[ENC]   parsing rule 6 RESERVED_BIT
Sep  1 08:22:35 03[ENC]   parsing rule 7 RESERVED_BIT
Sep  1 08:22:35 03[ENC]   parsing rule 8 FLAG
Sep  1 08:22:35 03[ENC]   parsing rule 9 FLAG
Sep  1 08:22:35 03[ENC]   parsing rule 10 FLAG
Sep  1 08:22:35 03[ENC]   parsing rule 11 FLAG
Sep  1 08:22:35 03[ENC]   parsing rule 12 FLAG
Sep  1 08:22:35 03[ENC]   parsing rule 13 FLAG
Sep  1 08:22:35 03[ENC]   parsing rule 14 U_INT_32
Sep  1 08:22:35 03[ENC]   parsing rule 15 HEADER_LENGTH
Sep  1 08:22:35 03[ENC] parsing HEADER payload finished
Sep  1 08:22:35 03[ENC] parsed a IKE_SA_INIT response header
Sep  1 08:22:35 03[NET] waiting for data on sockets
Sep  1 08:22:35 05[MGR] checkout IKEv2 SA by message with SPIs 482ecf21d8a833a8_i 79c465843445faa7_r
Sep  1 08:22:35 05[MGR] IKE_SA connectionAB[3] successfully checked out
Sep  1 08:22:35 05[NET] <connectionAB|3> received packet: from B[4500] to A[4500] (280 bytes)
Sep  1 08:22:35 11[MGR] <connectionAB|3> checkin IKE_SA connectionAB[3]
Sep  1 08:22:35 11[MGR] <connectionAB|3> checkin of IKE_SA successful

**Reply by tobiasbrunner** [reply]
The other peer apparently deletes the new IKE_SA (`Sep 1 08:22:35 10[IKE] <connectionAB|2> received DELETE for IKE_SA connectionAB[2]`). So check the log there.
